expr/func_linearregression.go

```diff
@@ -0,0 +1,142 @@
+			},
+			ArgString{
+				key:       "startSourceAt",
+				opt:       true,
+				validator: []Validator{IsATTime},
+				val:       &s.startSourceAt,
+			},
+			ArgString{
+				key:       "endSourceAt",
+				opt:       true,
+				validator: []Validator{IsATTime},
+				val:       &s.endSourceAt,
+			},
+		},
+		[]Arg{ArgSeriesList{}}
+}
+
+func (s *FuncLinearRegression) Context(context Context) Context {
+	s.startTargetAt = context.from
+	s.endTargetAt = context.to
+
+	return context
+}
+
+func linearRegressionAnalysis(series models.Series, startSourceAt uint32, endSourceAt uint32) (float64, float64, bool) {
+	var n float64
+	var sumI float64
+	var sumII float64
+	var sumV float64
+	var sumIV float64
+
+	for i := sort.Search(len(series.Datapoints), func(i int) bool { return series.Datapoints[i].Ts >= startSourceAt }); i < len(series.Datapoints) && series.Datapoints[i].Ts <= endSourceAt; i++ {
+		// The index must be rebuilt from the timestamp because
+		// the points of the series do not include "missing" points.
+		index := float64((series.Datapoints[i].Ts - startSourceAt) / series.Interval)
+		value := series.Datapoints[i].Val
+
+		n++
+		sumI += index
+		sumII += index * index
+		sumV += value
+		sumIV += index * value
+	}
+
+	denominator := n*sumII - sumI*sumI
+	if denominator == 0 {
+		return 0, 0, false
+	}
+
+	factor := (n*sumIV - sumI*sumV) / denominator / float64(series.Interval)
+	offset := (sumII*sumV-sumIV*sumI)/denominator - factor*float64(startSourceAt)
+	return factor, offset, true
+}
+
+func (s *FuncLinearRegression) Exec(dataMap DataMap) ([]models.Series, error) {
+	loc, err := time.LoadLocation("") // todo, no idea what timezone to use here, utc is a reasonable default
+	if err != nil {
+		return nil, fmt.Errorf("failed to load location: %w", err)
+	}
+
+	now := time.Now()
+
+	defaultStartSourceAt := uint32(now.Add(-24 * time.Hour).Unix())
+	startSourceAt, err := dur.ParseDateTime(s.startSourceAt, loc, now, defaultStartSourceAt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse 'startSourceAt' argument %q: %w", s.startSourceAt, err)
+	}
+
+	defaultEndSourceAt := uint32(now.Unix())
+	endSourceAt, err := dur.ParseDateTime(s.endSourceAt, loc, now, defaultEndSourceAt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse 'endSourceAt' argument %q: %w", s.endSourceAt, err)
+	}
+
+	// todo update context.from and .to if needed?
+
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	results := []models.Series{}
+	for _, serie := range series {
+		factor, offset, forecast := linearRegressionAnalysis(serie, startSourceAt, endSourceAt)
+		if !forecast {
+			continue
+		}
+
+		datapoints := []schema.Point{}
+		{
+			var i uint32
+			for i = 0; i <= (s.endTargetAt-s.startTargetAt)/serie.Interval; i++ {
+				datapoints = append(datapoints, schema.Point{
+					Val: offset + (float64(s.startTargetAt)+float64(i)*float64(serie.Interval))*factor,
+					Ts:  s.startTargetAt + i*serie.Interval,
```

Comment:
I don't think this is correct. These timestamps need to be adjusted back to the range that the user requested. 

For example, if the user is graphing for the last hour but used the last 24h to compute the regression, this should still only return data for the last hour. Similarly, if the user used windows of `-24h, -23h` as start/end but had a window of last 1h in grafana, it should still be graphing for the last 1h.

Reply:
See how [timeShift](https://github.com/grafana/metrictank/blob/master/expr/func_timeshift.go#L18) remembers the shift seconds and offsets the timestamps

Reply:
> if the user is graphing for the last hour but used the last 24h to compute the regression, this should still only return data for the last hour

Is that not what this is doing though? In this case the last hour would be startTargetAt = now-1h and endTargetAt would be now

Reply:
Ah, I misread these as the `Source` versions.